pymilvus/milvus_client/milvus_client.py

```diff
@@ -0,0 +1,801 @@
+
+    def _create_index(self) -> None:
+        """Create a index on the collection"""
+        if self._get_index() is None:
+            # If no index params, use a default HNSW based one
+            if self.index_params is None:
+                # TODO: Once segment normalization we can default to IP
+                metric_type = (
+                    "L2"
+                    if self.fields[self.vector_field] == DataType.FLOAT_VECTOR
+                    else "JACCARD"
+                )
+                # TODO: Once AUTOINDEX type is supported by Milvus we can default to HNSW always
+                index_type = "HNSW" if self.is_self_hosted else "AUTOINDEX"
+                params = {"M": 8, "efConstruction": 64} if self.is_self_hosted else {}
+                self.index_params = {
+                    "metric_type": metric_type,
+                    "index_type": index_type,
+                    "params": params,
+                }
+            try:
+                self.collection.create_index(
+                    self.vector_field,
+                    index_params=self.index_params,
+                    using=self.alias,
+                    timeout=self.timeout,
+                )
+                logger.info(
+                    "Successfully created an index on collection: %s",
+                    self.collection_name,
+                )
+
+            except MilvusException as ex:
+                logger.error(
+                    "Failed to create an index on collection: %s", self.collection_name
+                )
+                raise ex
+        else:
+            logger.debug(
+                "Index exists already for collection: %s", self.collection_name
+            )
+
+    def _get_index(self):
+        """Return the index dict if index exists."""
+        for index in self.collection.indexes:
+            if index.field_name == self.vector_field:
+                return index
+        return None
+
+    def _create_default_search_params(self) -> None:
+        """Generate search params based on the current index type"""
+        index = self._get_index().to_dict()
+        if index is not None:
+            index_type = index["index_param"]["index_type"]
+            metric_type = index["index_param"]["metric_type"]
+            self.default_search_params = DEFAULT_SEARCH_PARAMS[index_type]
+            self.default_search_params["metric_type"] = metric_type
+
+    def _load(self):
+        """Loads the collection."""
+        if self._get_index() is not None:
+            # Check if the collection is loaded or in progress of loading
+            if (
+                utility.load_state(
+                    self.collection_name, using=self.alias, timeout=self.timeout
+                )
+                != LoadState.NotLoad
+            ):
+                # IF the collection is loaded/loading, check the replica count
+                if len(self.collection.get_replicas().groups) == self.replica_number:
```

Comment:
pls open a issue in Milvus.
We have to make load idempotent.
It should dynamic add or remove one replica when ever you call load

Reply:
I though there is a refresh parameter which did this, pleas double check with Alex @soothing-rain

Reply:
We have this for refreshing. The description didnt sound like that logic but I could be wrong:
<img width="679" alt="image" src="https://user-images.githubusercontent.com/81822489/234070908-b771b90d-5408-4f4d-a7f2-4a513a7fb42e.png">

Reply:
I think if load is done. calling another load should just return success if collection is already loaded.

Reply:
I believe that it is idempotent, it was more for the dynamic replica count that I did this logic.